File: iOS/Sources/AppDimens/AppDimens.swift
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/bodenberg/appdimens.git
 * Date: 2025-01-15
 *
 * Library: AppDimens iOS - Main Export
 *
 * Description:
 * Main export file for the AppDimens library, providing access to all
 * functionality across all modules (Core, UI, and Games).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import UIKit

// MARK: - UI Module (if available)
#if canImport(SwiftUI)
@_exported import AppDimensUI
#endif

// MARK: - Games Module (if Metal is available)
#if canImport(Metal)
@_exported import AppDimensGames
#endif

// MARK: - Main AppDimens Class

/**
 * [EN] Main AppDimens class that provides access to fixed and dynamic dimension builders.
 * [PT] Classe principal AppDimens que fornece acesso aos construtores de dimensões fixas e dinâmicas.
 */
public class AppDimens {
    
    // MARK: - Singleton
    
    public static let shared = AppDimens()
    
    private init() {}
    
    // MARK: - Builder Methods
    
    /**
     * [EN] Creates a fixed dimension builder from a CGFloat value.
     * @param initialValue The initial base value.
     * @param ignoreMultiWindowAdjustment Whether to ignore multi-window adjustments.
     * @return An AppDimensFixed instance for chaining.
     * [PT] Cria um construtor de dimensão fixa a partir de um valor CGFloat.
     * @param initialValue O valor base inicial.
     * @param ignoreMultiWindowAdjustment Se deve ignorar ajustes multi-window.
     * @return Uma instância AppDimensFixed para encadeamento.
     */
    public func fixed(_ initialValue: CGFloat, ignoreMultiWindowAdjustment: Bool? = nil) -> AppDimensFixed {
        let ignore = ignoreMultiWindowAdjustment ?? false
        return AppDimensFixed(initialValue, ignoreMultiWindowAdjustment: ignore)
    }
    
    /**
     * [EN] Creates a fixed dimension builder from an Int value.
     * @param initialValue The initial base value.
     * @param ignoreMultiWindowAdjustment Whether to ignore multi-window adjustments.
     * @return An AppDimensFixed instance for chaining.
     * [PT] Cria um construtor de dimensão fixa a partir de um valor Int.
     * @param initialValue O valor base inicial.
     * @param ignoreMultiWindowAdjustment Se deve ignorar ajustes multi-window.
     * @return Uma instância AppDimensFixed para encadeamento.
     */
    public func fixed(_ initialValue: Int, ignoreMultiWindowAdjustment: Bool? = nil) -> AppDimensFixed {
        return fixed(CGFloat(initialValue), ignoreMultiWindowAdjustment: ignoreMultiWindowAdjustment)
    }
    
    /**
     * [EN] Creates a dynamic dimension builder from a CGFloat value.
     * @param initialValue The initial base value.
     * @param ignoreMultiWindowAdjustment Whether to ignore multi-window adjustments.
     * @return An AppDimensDynamic instance for chaining.
     * [PT] Cria um construtor de dimensão dinâmica a partir de um valor CGFloat.
     * @param initialValue O valor base inicial.
     * @param ignoreMultiWindowAdjustment Se deve ignorar ajustes multi-window.
     * @return Uma instância AppDimensDynamic para encadeamento.
     */
    public func dynamic(_ initialValue: CGFloat, ignoreMultiWindowAdjustment: Bool? = nil) -> AppDimensDynamic {
        let ignore = ignoreMultiWindowAdjustment ?? false
        return AppDimensDynamic(initialValue, ignoreMultiWindowAdjustment: ignore)
    }
    
    /**
     * [EN] Creates a dynamic dimension builder from an Int value.
     * @param initialValue The initial base value.
     * @param ignoreMultiWindowAdjustment Whether to ignore multi-window adjustments.
     * @return An AppDimensDynamic instance for chaining.
     * [PT] Cria um construtor de dimensão dinâmica a partir de um valor Int.
     * @param initialValue O valor base inicial.
     * @param ignoreMultiWindowAdjustment Se deve ignorar ajustes multi-window.
     * @return Uma instância AppDimensDynamic para encadeamento.
     */
    public func dynamic(_ initialValue: Int, ignoreMultiWindowAdjustment: Bool? = nil) -> AppDimensDynamic {
        return dynamic(CGFloat(initialValue), ignoreMultiWindowAdjustment: ignoreMultiWindowAdjustment)
    }
}

// MARK: - Global Cache Control

/**
 * [EN] Global cache control for all AppDimens instances.
 * [PT] Controle global de cache para todas as instâncias AppDimens.
 */
public class AppDimensGlobal {
    
    /**
     * [EN] Global cache control for all AppDimensDynamic instances.
     * [PT] Controle global de cache para todas as instâncias AppDimensDynamic.
     */
    public static var globalCacheEnabled: Bool = true {
        didSet {
            if !globalCacheEnabled {
                // Clear all caches when globally disabled
                clearAllCaches()
            }
        }
    }
    
    /**
     * [EN] Clears all caches from all instances.
     * [PT] Limpa todos os caches de todas as instâncias.
     */
    public static func clearAllCaches() {
        // This would need to be implemented with a registry of instances
        // For now, individual instances will clear their own caches
    }
}

// MARK: - Library Information

/**
 * [EN] Information about the AppDimens library.
 * [PT] Informações sobre a biblioteca AppDimens.
 */
public struct AppDimensInfo {
    public static let version = "1.0.5"
    public static let libraryName = "AppDimens"
    public static let description = "A responsive dimension management system for iOS"
    
    /**
     * [EN] Gets the library information as a dictionary.
     * @return A dictionary containing library information.
     * [PT] Obtém as informações da biblioteca como um dicionário.
     * @return Um dicionário contendo informações da biblioteca.
     */
    public static func info() -> [String: String] {
        var info: [String: String] = [
            "version": version,
            "libraryName": libraryName,
            "description": description
        ]
        
        // Add module information
        info["coreModule"] = AppDimensCoreInfo.moduleName
        info["coreVersion"] = AppDimensCoreInfo.version
        
        #if canImport(SwiftUI)
        info["uiModule"] = AppDimensUIInfo.moduleName
        info["uiVersion"] = AppDimensUIInfo.version
        #endif
        
        #if canImport(Metal)
        info["gamesModule"] = AppDimensGamesInfo.moduleName
        info["gamesVersion"] = AppDimensGamesInfo.version
        #endif
        
        return info
    }
    
    /**
     * [EN] Gets the available modules.
     * @return An array of available module names.
     * [PT] Obtém os módulos disponíveis.
     * @return Um array com os nomes dos módulos disponíveis.
     */
    public static func availableModules() -> [String] {
        var modules = ["Core"]
        
        #if canImport(SwiftUI)
        modules.append("UI")
        #endif
        
        #if canImport(Metal)
        modules.append("Games")
        #endif
        
        return modules
    }
}

File: ReactNative/src/core/AppDimens.ts
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/bodenberg/appdimens.git
 * Date: 2025-01-27
 *
 * Library: AppDimens React Native - Main Class
 *
 * Description:
 * Main AppDimens class providing access to fixed and dynamic dimension builders.
 * Acts as the primary gateway for responsive dimension management.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Dimensions, PixelRatio} from 'react-native';
import {AppDimensFixed} from './AppDimensFixed';
import {AppDimensDynamic} from './AppDimensDynamic';
import {AppDimensAdjustmentFactors} from './AppDimensAdjustmentFactors';
import {
  DeviceInfo,
  ScreenDimensions,
  BASE_WIDTH_DP,
  BASE_HEIGHT_DP,
} from '../types';

/**
 * Main AppDimens class providing access to dimension builders
 */
export class AppDimens {
  private static instance: AppDimens;
  private globalCacheEnabled: boolean = true;
  private globalAspectRatioEnabled: boolean = true;
  private globalIgnoreMultiWindowAdjustment: boolean = false;

  private constructor() {}

  /**
   * Get singleton instance
   */
  public static getInstance(): AppDimens {
    if (!AppDimens.instance) {
      AppDimens.instance = new AppDimens();
    }
    return AppDimens.instance;
  }

  /**
   * Create fixed dimension builder
   */
  public fixed(
    initialValue: number,
    ignoreMultiWindowAdjustment?: boolean,
  ): AppDimensFixed {
    const finalIgnoreMultiWindow =
      ignoreMultiWindowAdjustment ?? this.globalIgnoreMultiWindowAdjustment;
    const builder = new AppDimensFixed(initialValue, finalIgnoreMultiWindow);

    if (!this.globalAspectRatioEnabled) {
      builder.aspectRatio(false);
    }

    return builder;
  }

  /**
   * Create dynamic dimension builder
   */
  public dynamic(
    initialValue: number,
    ignoreMultiWindowAdjustment?: boolean,
  ): AppDimensDynamic {
    const finalIgnoreMultiWindow =
      ignoreMultiWindowAdjustment ?? this.globalIgnoreMultiWindowAdjustment;
    return new AppDimensDynamic(initialValue, finalIgnoreMultiWindow);
  }

  /**
   * Calculate percentage-based dimension
   */
  public percentage(
    percentage: number,
    type: 'lowest' | 'highest' = 'lowest',
  ): number {
    if (percentage < 0 || percentage > 1) {
      throw new Error('Percentage must be between 0.0 and 1.0');
    }

    const screenDimensions =
      AppDimensAdjustmentFactors.getCurrentScreenDimensions();
    const {width, height} = screenDimensions;

    const dimensionToUse =
      type === 'highest' ? Math.max(width, height) : Math.min(width, height);

    return dimensionToUse * percentage;
  }

  /**
   * [EN] Calculate percentage-based dimension in Dp.
   * Compatible with Android API.
   * [PT] Calcula dimensão baseada em porcentagem em Dp.
   * Compatível com API Android.
   */
  public dynamicPercentageDp(
    percentage: number,
    type: 'lowest' | 'highest' = 'lowest',
  ): number {
    return this.percentage(percentage, type);
  }

  /**
   * [EN] Calculate percentage-based dimension in physical pixels.
   * Compatible with Android API.
   * [PT] Calcula dimensão baseada em porcentagem em pixels físicos.
   * Compatível com API Android.
   */
  public dynamicPercentagePx(
    percentage: number,
    type: 'lowest' | 'highest' = 'lowest',
  ): number {
    const dpValue = this.dynamicPercentageDp(percentage, type);
    return AppDimensAdjustmentFactors.pointsToPixels(dpValue);
  }

  /**
   * [EN] Calculate percentage-based dimension in scalable pixels (Sp).
   * Compatible with Android API.
   * [PT] Calcula dimensão baseada em porcentagem em pixels escaláveis (Sp).
   * Compatível com API Android.
   */
  public dynamicPercentageSp(
    percentage: number,
    type: 'lowest' | 'highest' = 'lowest',
  ): number {
    const dpValue = this.dynamicPercentageDp(percentage, type);
    const fontScale = AppDimensAdjustmentFactors.getFontScale();
    return dpValue * fontScale;
  }

  /**
   * Calculate available item count in a container
   */
  public calculateAvailableItemCount(
    containerSize: number,
    itemSize: number,
    itemMargin: number,
  ): number {
    const totalItemSize = itemSize + itemMargin * 2;
    return totalItemSize > 0 ? Math.floor(containerSize / totalItemSize) : 0;
  }

  /**
   * Convert millimeters to pixels
   */
  public mm(value: number): number {
    // 1 inch = 25.4 mm
    // 1 inch = 160 dp (Android standard)
    // 1 dp = 1 pixel in React Native (approximately)
    return (value / 25.4) * 160;
  }

  /**
   * Convert centimeters to pixels
   */
  public cm(value: number): number {
    return this.mm(value * 10);
  }

  /**
   * Convert inches to pixels
   */
  public inch(value: number): number {
    return value * 160; // 160 dp per inch
  }

  /**
   * Get current device information
   */
  public getDeviceInfo(): DeviceInfo {
    return AppDimensAdjustmentFactors.getCurrentDeviceInfo();
  }

  /**
   * Get current screen dimensions
   */
  public getScreenDimensions(): ScreenDimensions {
    return AppDimensAdjustmentFactors.getCurrentScreenDimensions();
  }

  /**
   * Set global cache control
   */
  public setGlobalCache(enabled: boolean): AppDimens {
    this.globalCacheEnabled = enabled;
    if (!enabled) {
      AppDimensAdjustmentFactors.clearCache();
    }
    return this;
  }

  /**
   * Get global cache setting
   */
  public isGlobalCacheEnabled(): boolean {
    return this.globalCacheEnabled;
  }

  /**
   * Set global aspect ratio adjustment
   */
  public setGlobalAspectRatio(enabled: boolean): AppDimens {
    this.globalAspectRatioEnabled = enabled;
    return this;
  }

  /**
   * Get global aspect ratio setting
   */
  public isGlobalAspectRatioEnabled(): boolean {
    return this.globalAspectRatioEnabled;
  }

  /**
   * Set global multi-window adjustment
   */
  public setGlobalIgnoreMultiWindowAdjustment(ignore: boolean): AppDimens {
    this.globalIgnoreMultiWindowAdjustment = ignore;
    return this;
  }

  /**
   * Get global multi-window adjustment setting
   */
  public isGlobalIgnoreMultiWindowAdjustment(): boolean {
    return this.globalIgnoreMultiWindowAdjustment;
  }

  /**
   * Clear all caches
   */
  public clearAllCaches(): void {
    AppDimensAdjustmentFactors.clearCache();
  }

  /**
   * Get library information
   */
  public getInfo(): Record<string, string> {
    return {
      version: '1.0.7',
      libraryName: 'AppDimens React Native',
      description: 'Smart and responsive dimensioning for React Native',
      platform: 'React Native',
      author: 'Jean Bodenberg',
    };
  }

  /**
   * Get performance metrics
   */
  public getPerformanceMetrics(): {
    cacheSize: number;
    deviceInfo: DeviceInfo;
    screenDimensions: ScreenDimensions;
  } {
    const cacheStats = AppDimensAdjustmentFactors.getCacheStats();
    return {
      cacheSize: cacheStats.size,
      deviceInfo: this.getDeviceInfo(),
      screenDimensions: this.getScreenDimensions(),
    };
  }
}

// Export singleton instance
export const appDimens = AppDimens.getInstance();
